## Defining the question

### Specifying the data analytical question

Implement a naives bayes Classifier model, conduct different experiments, and calculate the resulting metrics

### Defining the Metric of success

Following performance metrics are used to analyze the 
performance of the Naive bayes classifier models with different metrics.

1. **Precision**: It is used to measures the classifier exactness. It is 
also known as positive predictive value (PPV). 

2. **Recall**: It measures the classifier completeness. It is also known 
as sensitivity or hit rate, or true positive rate (TPR). 
Specificity: It measures the proportion of actual negatives that 
are correctly identified and also called a true negative rate 
(TNR). 

3. **Accuracy**: It measures the fraction of predictions our model got 
right. It can be calculated as the total number of correct 
predictions divided by the total number of test cases. 

4. **F1-Score**: It measures the balance between precision and recall. 


### Understanding the context

Naive Bayes uses a similar method to predict the probability of different class based on various attributes. This algorithm is mostly used in text classification and with problems having multiple classes.

In this context to classifiy the an email whether it is spam or not, I shall employ gaussian Naive Bayes Classifier model and test different metrics on the model and observe it's performance. 

Gaussian Naive Bayes Classifier model shall be applied since the predictor data is continuos

### Recording experimental design

The following are the steps taken to achieve the set objectives:
1. Importing the relevant libraries
2. Loading dataset
3. Previewing the dataset
4. Tidying the dataset- this will entail steps such as checking validity of the attributes in the dataset, check for any missing values in the dataset, check and deal with duplicates appropiately in the dataset
5. Implementing the solution - Here four different models shall be created using 80%, 70%, 60% and Finally 20% of the data to train the model and observing the performance metrics of each classifier
6. Optimization of the model with the best performance metrics
7. Challenging the naives bayes model using a random forest classifier
8. Drawing conclusions
9. Follow up questions

### Data Relevance

## Importing libraries

In [ ]:
# Importing pandas
import pandas as pd

# Importing numpy
import numpy as np

# Importing matplotlib
import matplotlib.pyplot as plt

# Importing seaborn
import seaborn as sns

# Importing datetime
import datetime

# Importing scipy.stats
import scipy.stats as stats

# Importing scipy
import scipy as sp

# Importing labelencoder from sklearn preprocessing
from sklearn.preprocessing import LabelEncoder

## Loading dataset

In [ ]:
# Loading the spam dataset
data =  pd.read_csv('/content/spambase.data', sep=",", header = None)


## Previewing dataset

In [ ]:
# Previewing the first five records of the spam data frame
data.head()

0     1     2    3     4     5     6     7     8     9   ...    48  \
0  0.00  0.64  0.64  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.00   
1  0.21  0.28  0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
2  0.06  0.00  0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
3  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   

      49   50     51     52     53     54   55    56  57  
0  0.000  0.0  0.778  0.000  0.000  3.756   61   278   1  
1  0.132  0.0  0.372  0.180  0.048  5.114  101  1028   1  
2  0.143  0.0  0.276  0.184  0.010  9.821  485  2259   1  
3  0.137  0.0  0.137  0.000  0.000  3.537   40   191   1  
4  0.135  0.0  0.135  0.000  0.000  3.537   40   191   1  

[5 rows x 58 columns]

In [ ]:
# Previewing last five records of the spam dataset
data.tail()

0    1     2    3     4     5    6    7    8    9   ...     48     49  \
4596  0.31  0.0  0.62  0.0  0.00  0.31  0.0  0.0  0.0  0.0  ...  0.000  0.232   
4597  0.00  0.0  0.00  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000  0.000   
4598  0.30  0.0  0.30  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.102  0.718   
4599  0.96  0.0  0.00  0.0  0.32  0.00  0.0  0.0  0.0  0.0  ...  0.000  0.057   
4600  0.00  0.0  0.65  0.0  0.00  0.00  0.0  0.0  0.0  0.0  ...  0.000  0.000   

       50     51   52   53     54  55   56  57  
4596  0.0  0.000  0.0  0.0  1.142   3   88   0  
4597  0.0  0.353  0.0  0.0  1.555   4   14   0  
4598  0.0  0.000  0.0  0.0  1.404   6  118   0  
4599  0.0  0.000  0.0  0.0  1.147   5   78   0  
4600  0.0  0.125  0.0  0.0  1.250   5   40   0  

[5 rows x 58 columns]

In [ ]:
# Previewing column information of the attributes in the spam dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4601 non-null   float64
 1   1       4601 non-null   float64
 2   2       4601 non-null   float64
 3   3       4601 non-null   float64
 4   4       4601 non-null   float64
 5   5       4601 non-null   float64
 6   6       4601 non-null   float64
 7   7       4601 non-null   float64
 8   8       4601 non-null   float64
 9   9       4601 non-null   float64
 10  10      4601 non-null   float64
 11  11      4601 non-null   float64
 12  12      4601 non-null   float64
 13  13      4601 non-null   float64
 14  14      4601 non-null   float64
 15  15      4601 non-null   float64
 16  16      4601 non-null   float64
 17  17      4601 non-null   float64
 18  18      4601 non-null   float64
 19  19      4601 non-null   float64
 20  20      4601 non-null   float64
 21  21      4601 non-null   float64
 22  

All columns have appropiate data types

In [ ]:
# Previewing the shape(rows and columns) of the training dataset

rows, columns = data.shape

# Printing out the number of rows and columns in the training dataset
print(f'The training dataset has {rows} rows, and {columns} columns')

The training dataset has 4601 rows, and 58 columns


## Tidying dataset

### Validity

In [ ]:
# Checking the column names in the spam dataset
data.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57],
           dtype='int64')

In [ ]:
# Renaming column 57(the target variable) in the spam dataset appropiately
data.rename(columns = {57: 'spam or not spam'}, inplace = True)

In [ ]:
# Previewing column names after renaming the target variable in spam dataset appropiately
data.columns

Index([                 0,                  1,                  2,
                        3,                  4,                  5,
                        6,                  7,                  8,
                        9,                 10,                 11,
                       12,                 13,                 14,
                       15,                 16,                 17,
                       18,                 19,                 20,
                       21,                 22,                 23,
                       24,                 25,                 26,
                       27,                 28,                 29,
                       30,                 31,                 32,
                       33,                 34,                 35,
                       36,                 37,                 38,
                       39,                 40,                 41,
                       42,                 43,                

Columns have been appropiate data types for the analysis

### Missing values

In [ ]:
# Checking for anly missing values in the spam dataset
data.isnull().any().sum()

0

The dataset does not have any missing values

### Duplicates

In [ ]:
# Checking for any duplicate records in the spam dataset
print(f' The spam dataset has {data.duplicated().sum()} duplicate records')

 The spam dataset has 391 duplicate records


In [ ]:
# Select duplicate rows except first occurrence based on all columns in the spam dataset
duplicates_df = data[data.duplicated()]

# Print duplicates dataframe first five records
duplicates_df.head()

0    1     2    3    4     5    6    7    8    9  ...     48     49  \
26   0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0  0.0  0.0  ...  0.000  0.196   
103  0.0  0.0  0.64  0.0  0.0  0.64  0.0  0.0  0.0  0.0  ...  0.094  0.189   
104  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0  0.0  0.0  ...  0.305  0.611   
105  0.0  0.0  0.64  0.0  0.0  0.64  0.0  0.0  0.0  0.0  ...  0.094  0.189   
106  0.0  0.0  0.00  0.0  0.0  0.00  0.0  0.0  0.0  0.0  ...  0.305  0.611   

        50     51     52   53      54   55   56  spam or not spam  
26   0.000  0.392  0.196  0.0   5.466   22   82                 1  
103  0.284  0.662  0.000  0.0  10.068  131  292                 1  
104  0.000  1.529  0.000  0.0   5.500   22   66                 1  
105  0.284  0.662  0.000  0.0  10.068  131  292                 1  
106  0.000  1.529  0.000  0.0   5.500   22   66                 1  

[5 rows x 58 columns]

In [ ]:
# Preview the last five records of the duplicates dataframe
duplicates_df.tail()

0    1     2    3    4    5    6    7    8    9  ...     48     49  \
4439  0.0  0.0  0.74  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.036  0.147   
4441  0.0  0.0  0.74  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.036  0.147   
4537  0.0  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000  0.000   
4541  0.0  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000  0.000   
4550  0.0  0.0  0.00  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.000  0.000   

       50     51   52   53     54  55   56  spam or not spam  
4439  0.0  0.147  0.0  0.0  2.587  55  282                 0  
4441  0.0  0.147  0.0  0.0  2.587  55  282                 0  
4537  0.0  0.000  0.0  0.0  1.000   1    6                 0  
4541  0.0  0.000  0.0  0.0  1.000   1    2                 0  
4550  0.0  0.000  0.0  0.0  2.000   3    4                 0  

[5 rows x 58 columns]

In [ ]:
# Dropping duplicate records in the spam dataset
spam_df = data.drop_duplicates()

In [ ]:
# Check for any remaining duplicate records in the spam dataset
spam_df.duplicated().sum()

0

In [ ]:
# Preview records lost after dropping duplicates
print(f'The number of records lost after dropping duplicates was: {(len(data)- len(spam_df))}')

The number of records lost after dropping duplicates was: 391


## Implementing the solution

To predict whether an email is spam or not, the spam dataframe shall be split into the a train and test set. A Gaussian Naive Bayes Classifier algorithim shall be used since the predictors are continuos in nature. The predictors are assumed to be normally distributed.

The train set shall be be used to train the Gaussian Naive Bayes Classifier model using first 80% of data as train 20% as test, the 70% of data as train and 30% as test and finally 60% of data as train and 40% as test

The model shall be optimized and the metrics of optimized model analysed


### Model 1

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 80% of the data the test set is going to take up 20%

In [ ]:
# Splitting the spam dataframe to get the target variables and the feature variables
# Feature variables
X = spam_df.drop(columns = ['spam or not spam'], axis =1)

# Target varibale
y = spam_df['spam or not spam']

In [ ]:
# Importing train test split from sklearn model selection library
from sklearn.model_selection import train_test_split

In [ ]:
# Splitting the target and feature variables  to get 80% of the data used in training the 
# Model and 20% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [ ]:
# Importing GaussianNB from sklearn naive bayes library
from sklearn.naive_bayes import GaussianNB

In [ ]:
# Defining the classifier model
clf1 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf1.fit(X_train, y_train) 

In [ ]:
# Making predictions from the test set of the spam dataset
#
pred = clf1.predict(X_test)

In [ ]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[359 158]
 [ 10 315]]


Confusion matrix interpretation:

1. True positive = 362
2. True Negative = 320
3. False positive = 16
4. False negative = 144

In [ ]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.69      0.81       517
           1       0.67      0.97      0.79       325

    accuracy                           0.80       842
   macro avg       0.82      0.83      0.80       842
weighted avg       0.85      0.80      0.80       842



The model (Model 1) had an accuracy score of 0.80

Using 80% of the data for the training set, the model (model 1) performed very well. It corretly predicted none spam emails 97/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 67 out 100 times correctly.

Model1 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

### Model 2

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 70% of the data the test set is going to take up 30%

In [ ]:
# Splitting the target and feature variables  to get 70% of the data used in training the 
# Model and 30% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [ ]:
# Defining the classifier model
clf2 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf2.fit(X_train, y_train) 

In [ ]:
# Making predictions from the test set of the spam dataset
#
pred = clf2.predict(X_test)

In [ ]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[577 208]
 [ 15 463]]


Confusion matrix interpretation:

1. True positive = 562
2. True Negative = 479
3. False positive = 18
4. False negative = 204

In [ ]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.74      0.84       785
           1       0.69      0.97      0.81       478

    accuracy                           0.82      1263
   macro avg       0.83      0.85      0.82      1263
weighted avg       0.87      0.82      0.83      1263



The model (Model 2) had an accuracy score of 0.82

Using 70% of the data for the training set, the model (model 2) performed very well. It corretly predicted none spam emails 97/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 69 out 100 times correctly.

Model2 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

However the model(model 2) generally produced better results relative to the first model(model 1)

### Model 3

In [ ]:
# Splitting the target and feature variables  to get 60% of the data used in training the 
# Model and 40% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40)

In [ ]:
# Defining the classifier model
clf3 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf3.fit(X_train, y_train) 

In [ ]:
# Making predictions from the test set of the spam dataset
#
pred = clf2.predict(X_test)

In [ ]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[726 255]
 [ 30 673]]


Confusion matrix interpretation:

1. True positive = 775
2. True Negative = 604
3. False positive = 30
4. False negative = 275

In [ ]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.96      0.74      0.84       981
           1       0.73      0.96      0.83       703

    accuracy                           0.83      1684
   macro avg       0.84      0.85      0.83      1684
weighted avg       0.86      0.83      0.83      1684



The model (Model 3) had an accuracy score of 0.83

Using 60% of the data for the training set, the model (model 2) performed very well. It corretly predicted none spam emails 96/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 73 out 100 times correctly.

Model3 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

Model 3 had better performance metrics than than all the other two models. The performance metrics improve the more the training dataset is reduced in size

## Optimization

### Normalization of predictors

In creating the three models, I assumed the predictor data had a gaussin distribution.

In this step the predictor data shall be normalized before is it is fitted into the best performing model (model 2) that used 70% of the data as the training set

In [ ]:
# Importing MinMaxScaler from sklearn preprocessing library
from sklearn.preprocessing import MinMaxScaler

In [ ]:
# Splitting the target and feature variables  to get 60% of the data used in training the 
# Model and 40% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40)

In [ ]:
# Normalization of the predictor data
# The feature variables are normalized to ensure that they are all in a normally distributed

# Specifying the normalizer
norm = MinMaxScaler()

# Fitting the X train data to the scaler
norm.fit(X_train)

MinMaxScaler()

In [ ]:
# Normalization of the x train data
X_train = norm.transform(X_train)

# Normalization  of th x_test data
X_test = norm.transform(X_test)

In [ ]:
# Defining the classifier model
clf3 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf3.fit(X_train, y_train) 

In [ ]:
# Making predictions from the test set of the spam dataset
#
pred = clf3.predict(X_test)

In [ ]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[750 270]
 [ 27 637]]


Confusion matrix interpretation:

1. True positive = 750
2. True Negative = 270
3. False positive = 27
4. False negative = 270

In [ ]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.74      0.83      1020
           1       0.70      0.96      0.81       664

    accuracy                           0.82      1684
   macro avg       0.83      0.85      0.82      1684
weighted avg       0.86      0.82      0.83      1684



There was a slight improvement in the models performance metrics. The accuracy slightly decreased from 0.83 to 0.82.

This implies that the data distribution slightly affects the models performance metrics 

## Conclusion

## Follow up questions